core/src/main/clojure/xtdb/operator/window.clj

```diff
@@ -0,0 +1,237 @@
+                                         vw/->writer)
+              ^TreeSet min-max-map (TreeSet.)]
+          (reify
+            IWindowFnSpec
+            (aggregate [_ group-mapping sortMapping in-rel]
+              (let [in-vec (.readerForName in-rel (str from-name))
+                    row-count (.getValueCount group-mapping)]
+                (loop [previous-group nil idx 0]
+                  (when (< idx row-count)
+                    (let [sort-idx (aget sortMapping idx)
+                          group-idx (.get group-mapping sort-idx)]
+                      (when (not= group-idx previous-group)
+                        (.clear min-max-map))
+                      (.add min-max-map (.getObject in-vec sort-idx))
+                      (.writeObject out-vec (if (= op :min) (.first min-max-map) (.last min-max-map)))
+                      (recur group-idx (inc idx)))))))
+
+            (finish [_]
+              (vw/vec-wtr->rdr out-vec))
+
+            Closeable
+            (close [_] (.close out-vec))))))))
+
+
+
+(deftype WindowFnCursor [^BufferAllocator allocator
+                         ^ICursor in-cursor
+                         ^IPersistentMap static-fields
+                         ^IGroupMapper group-mapper
+                         order-specs
+                         ^List window-specs
+                         ^:unsynchronized-mutable ^boolean done?]
+  ICursor
+  (tryAdvance [this c]
+    (boolean
+     (when-not done?
+       (set! (.done? this) true)
+
+       (try
+         (let [grouping-vec-name "grouping-vec"]
+           ;; TODO we likely want to do some retaining here instead of copying
+           (util/with-open [rel-wtr (RelationWriter. allocator (for [^Field field static-fields]
+                                                                 (vw/->writer (.createVector field allocator))))
+                            ;; TODO check why grouping can be null
+                            ^IVectorWriter grouping-wtr (-> #xt.arrow/field ["grouping-vec" #xt.arrow/field-type [#xt.arrow/type :i32 true]]
+                                                            (.createVector allocator)
+                                                            vw/->writer)]
+
+             (.forEachRemaining in-cursor (fn [in-rel]
+                                            (vw/append-rel rel-wtr in-rel)
+                                            (with-open [group-mapping (.groupMapping group-mapper in-rel)]
+                                              (vw/append-vec grouping-wtr (vr/vec->reader group-mapping)))))
+
+             (let [rel-rdr (vw/rel-wtr->rdr rel-wtr)
+                   sort-mapping (order-by/sorted-idxs (vr/rel-reader (conj (seq rel-rdr) (vw/vec-wtr->rdr grouping-wtr)))
+                                                      (into [[(symbol grouping-vec-name)]] order-specs))]
+               (doseq [^IWindowFnSpec window-spec window-specs]
```

Comment:
given we're doing this all in one pass (cf. group-by where we call aggregate once for each input block) it might be simpler to combine aggregate and finish in the window operator?